org.eclipse.lyo.oslc4j.codegenerator/src/org/eclipse/lyo/oslc4j/codegenerator/files/generateServiceProviderCatalogSingleton.mtl

```diff
@@ -160,60 +160,39 @@ public class [javaClassNameForSingleton(aServiceProviderCatalog) /]
     }
 
-    [for (aServiceProvider: ServiceProvider | aServiceProviderCatalog.serviceProviders)]
-    public static boolean contains[aServiceProvider.javaName(true)/](final [javaClassName(aServiceProvider) /] serviceProviderInfo) {
-        return serviceProviders.containsKey([javaClassNameForFactory(aServiceProvider)/].constructIdentifier(serviceProviderInfo));
-    }
-
-    public static ServiceProvider get[aServiceProvider.javaName(true)/](HttpServletRequest httpServletRequest, [aServiceProvider.instanceMethodSignature()/])
-    {
-        ServiceProvider serviceProvider;
-
-        synchronized(serviceProviders)
-        {
-            String identifier = [javaClassNameForFactory(aServiceProvider)/].constructIdentifier([aServiceProvider.instanceMethodParameterList()/]);
-            serviceProvider = serviceProviders.get(identifier);
-
-            //One retry refreshing the service providers
-            if (serviceProvider == null)
-            {
-                getServiceProviders(httpServletRequest);
-                serviceProvider = serviceProviders.get(identifier);
-            }
-        }
-
-        if (serviceProvider != null)
-        {
-            return serviceProvider;
-        }
-
-        throw new WebApplicationException(Status.NOT_FOUND);
+    public static boolean containsServiceProvider(final String identifier) {
+        return serviceProviders.containsKey(identifier);
     }
 
-    public static ServiceProvider create[aServiceProvider.javaName(true)/](final [javaClassName(aServiceProvider) /] serviceProviderInfo) 
-            throws OslcCoreApplicationException, URISyntaxException, IllegalArgumentException {
-        String identifier = [javaClassNameForFactory(aServiceProvider)/].constructIdentifier(serviceProviderInfo);
-        if (contains[aServiceProvider.javaName(true)/](serviceProviderInfo)) {
-            throw new IllegalArgumentException(String.format("The SP '%s' was already registered", identifier));
-        }
-        return [javaClassNameForFactory(aServiceProvider)/].createServiceProvider(serviceProviderInfo);
+    public static boolean containsServiceProvider(final ServiceProvider serviceProvider) {
+        return containsServiceProvider(serviceProvider.getIdentifier());
     }
 
-    public static ServiceProvider register[aServiceProvider.javaName(true)/](final HttpServletRequest httpServletRequest,
+    public static ServiceProvider register(final HttpServletRequest httpServletRequest,
                                                           final ServiceProvider serviceProvider)
                                                 throws URISyntaxException
     {
+        return register(serviceProvider);
+    }
+
+    // This version is for self-registration and thus package-protected
+    static ServiceProvider register(final ServiceProvider serviceProvider)
+                                            throws URISyntaxException
+    {
+        if (containsServiceProvider(serviceProvider)) {
```

Comment:
This is still not thread-safe, we need the same check duplicated inside the `synchronized` block. Just noticed this.  See https://stackoverflow.com/questions/18093735/double-checked-locking-in-singleton

Reply:
Can you please check I did what you expected? I trust you on this, yet it was odd to me.
Main concern is: Can we safely raise an excption within a synchronised block?

Reply:
From Java Lang Spec  § 14.19:

>    If execution of the Block completes normally, then the lock is unlocked and the synchronized statement completes normally. If execution of the Block completes abruptly for any reason, then the lock is unlocked and the synchronized statement then completes abruptly for the same reason.